batch.go

```diff
@@ -522,3 +563,225 @@ func (c *cronTicker) Stop() {
 func (c *cronTicker) Next(now time.Time) time.Time {
 	return c.expr.Next(now)
 }
+
+// FluxQueryNode is a node for making flux queries
+type FluxQueryNode struct {
+	node
+	b        *pipeline.QueryFluxNode
+	query    *QueryFlux
+	ticker   ticker
+	queryMu  sync.Mutex
+	queryErr chan error
+	closing  chan struct{}
+	aborting chan struct{}
+
+	batchesQueried *expvar.Int
+	pointsQueried  *expvar.Int
+	byName         bool
+}
+
+func newQueryFluxNode(et *ExecutingTask, n *pipeline.QueryFluxNode, d NodeDiagnostic) (*FluxQueryNode, error) {
+	bn := &FluxQueryNode{
+		node:     node{Node: n, et: et, diag: d},
+		b:        n,
+		closing:  make(chan struct{}),
+		aborting: make(chan struct{}),
+	}
+	bn.node.runF = bn.runBatch
+	bn.node.stopF = bn.stopBatch
+
+	// Create query
+	q, err := NewQueryFlux(n.QueryStr, n.Org, n.OrgID)
+	if err != nil {
+		return nil, err
+	}
+	bn.query = q
+	// Determine schedule
+	if n.Every != 0 && n.Cron != "" {
+		return nil, errors.New("must not set both 'every' and 'cron' properties")
+	}
+	switch {
+	case n.Every > 0:
+		bn.ticker = newTimeTicker(n.Every, n.AlignFlag)
+	case n.Cron != "":
+		var err error
+		bn.ticker, err = newCronTicker(n.Cron)
+		if err != nil {
+			return nil, err
+		}
+	case n.Every < 0:
+		return nil, errors.New("'every' duration must must non-negative")
+	default:
+		return nil, errors.New("must define one of 'every' or 'cron'")
+	}
+
+	return bn, nil
+}
+
+func (n *FluxQueryNode) Start() {
+	n.queryMu.Lock()
+	defer n.queryMu.Unlock()
+	n.queryErr = make(chan error, 1)
+	go func() {
+		n.queryErr <- n.doQuery(n.ins[0])
+	}()
+}
+
+func (n *FluxQueryNode) Abort() {
+	close(n.aborting)
+}
+
+func (n *FluxQueryNode) Cluster() string {
+	return n.b.Cluster
+}
+
+func (n *FluxQueryNode) Queries(start, stop time.Time) ([]*QueryFlux, error) {
+	now := time.Now()
+	if stop.IsZero() {
+		stop = now
+	}
+	// Crons are sensitive to timezones.
+	// Make sure we are using local time.
+	current := start.Local()
+	queries := make([]*QueryFlux, 0)
+	for {
+		current = n.ticker.Next(current)
+		if current.IsZero() || current.After(stop) {
+			break
+		}
+		qstop := current.Add(-1 * n.b.Offset)
+		if qstop.After(now) {
+			break
+		}
+
+		q, err := n.query.Clone()
+		if err != nil {
+			return nil, err
+		}
+		q.Now = time.Now()
+		queries = append(queries, q)
+	}
+	return queries, nil
+}
+
+// Query InfluxDB and collect batches on batch collector.
+func (n *FluxQueryNode) doQuery(in edge.Edge) (err error) {
+	defer in.Close()
+	n.batchesQueried = &expvar.Int{}
+	n.pointsQueried = &expvar.Int{}
+
+	n.statMap.Set(statsBatchesQueried, n.batchesQueried)
+	n.statMap.Set(statsPointsQueried, n.pointsQueried)
+
+	if n.et.tm.InfluxDBService == nil {
+		return errors.New("InfluxDB not configured, cannot query InfluxDB for batch query")
+	}
+
+	con, err := n.et.tm.InfluxDBService.NewNamedClient(n.b.Cluster)
```

Comment:
Should we use the internal execution engine that tasks uses, or stick with sending to a remote influxdb?

Reply:
Answer: yes we should, but let's merge this first.